### PR TITLE
Make CA classes implement JSONSerializer to reduce boilerplate code

### PR DESCRIPTION
**Pull request by ckelleyRH**

### CODE REVIEW

**Comment by ckelleyRH** [comment]
There were no unit tests here either :-D

File: base/ca/src/main/java/org/dogtagpki/ct/CTEngine.java [hunk 1]
```diff
@@ -197,8 +198,8 @@ public void process(X509CertInfo certi, CertificateAuthority ctCA, AuthorityID a
 
                 /* TODO this should be a configurable; hardcoded for now */
                 boolean allowFailedSCTVerification = true;
-
-                boolean verified = verifySCT(CTResponse.fromJSON(respS), tbsCert, ls.getPublicKey(), ctCA);
+                final CTResponse response = JSONSerializer.fromJSON(respS, CTResponse.class);
```
[the remark this answers is not in the document]
Comment (edewata) [reply base/ca/src/main/java/org/dogtagpki/ct/CTEngine.java:201]: Hmm.. this is actually not much different from:
```
final CTResponse response = new ObjectMapper().readValue(respS, CTResponse.class);
```
Should we go ahead with the change?

[the remark this answers is not in the document]
Comment (ckelleyRH) [reply base/ca/src/main/java/org/dogtagpki/ct/CTEngine.java:201]: Final is not necessary, I'll remove it.

Yeah that's true it's not much different. Even the `toJSON()` method is probably not necessary. The only advantage, aside from already mentioned advantages, is that you get some consistency if you use `{to,fromJSON}`, use of `ObjectMapper` varies a bit around the code, but there's nothing preventing someone from ignoring/not noticing the new interface and using `ObjectMapper` anyway so I guess that's actually irrelevant.

[the remark this answers is not in the document]
Comment (edewata) [reply base/ca/src/main/java/org/dogtagpki/ct/CTEngine.java:201]: Does it make sense to keep the `toJSON()` and `fromJSON()` in all classes for usage simplicity, but modify them to call `JSONSerializer.toJSON()` and `JSONSerializer.fromJSON()` for consistency?

**Review by ckelleyRH — COMMENTED** [review]
If we did that there would be no need for `JSONSerializer` to be an interface, it could just be a utility class with some static methods in it, which I guess is fine.

I'm not sure. The original purpose was to reduce the amount of boilerplate code by extracting the duplicated serialisation code into a single location, keeping the methods _and_ introducing `JSONSerializer` feels wrong. You've still got the 90+ very similar methods but now also have an additional indirection layer.

**Comment by edewata** [comment]
If we use `ObjectMapper` directly and if we later need to modify something (e.g. configure `ObjectMapper` object) we might have to make the same changes in all usages, which there could be many more than the duplicate `toJSON()/fromJSON()` methods. If we encapsulate the code into `JSONSerializer` we only need to modify it in a single location.

So there's no absolute right or wrong here, but it's a matter of balancing the risk and benefits. Sorry, couldn't be more vague than this :) I'll let you decide. I'll ACK this PR anyway. Thanks!

**Review by edewata — APPROVED** [review]

**Comment by ckelleyRH** [comment]
I think I'll go with it as-is then, hopefully the fact that there will only be `ObjectMapper` refs in the new interface will encourage people to follow the pattern going forward.

**Pull request merged**